core/src/main/scala/kafka/server/metadata/RaftMetadataCache.scala

```diff
@@ -0,0 +1,389 @@
+  // if the leader is known but corresponding node with the listener name is not available, return Some(NO_NODE)
+  override def getPartitionLeaderEndpoint(topic: String, partitionId: Int, listenerName: ListenerName): Option[Node] = {
+    val image = _currentImage
+    image.partitions.topicPartition(topic, partitionId).map { partition =>
+      image.aliveBroker(partition.leaderId) match {
+        case Some(broker) =>
+          broker.endpoints.getOrElse(listenerName.value(), Node.noNode)
+        case None =>
+          Node.noNode
+      }
+    }
+  }
+
+  override def getPartitionReplicaEndpoints(tp: TopicPartition, listenerName: ListenerName): Map[Int, Node] = {
+    val image = _currentImage
+    image.partitions.topicPartition(tp.topic(), tp.partition()).map { partition =>
+      partition.replicas.asScala.map(replicaId => replicaId.intValue() -> {
+        image.aliveBroker(replicaId) match {
+          case Some(broker) =>
+            broker.endpoints.getOrElse(listenerName.value(), Node.noNode())
+          case None =>
+            Node.noNode()
+        }}).toMap
+        .filter(pair => pair match {
+          case (_, node) => !node.isEmpty
+        })
+    }.getOrElse(Map.empty[Int, Node])
+  }
+
+  override def getControllerId: Option[Int] = {
+    _currentImage.controllerId
+  }
+
+  override def getClusterMetadata(clusterId: String, listenerName: ListenerName): Cluster = {
+    val image = _currentImage
+    val nodes = new util.HashMap[Integer, Node]
+    image.brokers.aliveBrokers().foreach { node => if (!node.fenced) {
+      node.endpoints.get(listenerName.value()).foreach { nodes.put(node.id, _) }
+    }
+    }
+
+    def node(id: Integer): Node = {
+      Option(nodes.get(id)).getOrElse(new Node(id, "", -1))
+    }
+
+    val partitionInfos = new util.ArrayList[PartitionInfo]
+    val internalTopics = new util.HashSet[String]
+
+    image.partitions.allPartitions().foreach { partition =>
+      partitionInfos.add(new PartitionInfo(partition.topicName,
+        partition.partitionIndex, node(partition.leaderId),
+        partition.replicas.asScala.map(node).toArray,
+        partition.isr.asScala.map(node).toArray,
+        partition.offlineReplicas.asScala.map(node).toArray))
+      if (Topic.isInternal(partition.topicName)) {
+        internalTopics.add(partition.topicName)
+      }
+    }
+
+    new Cluster(clusterId, nodes.values(),
+      partitionInfos, Collections.emptySet[String], internalTopics,
+      node(Integer.valueOf(image.controllerId.getOrElse(-1))))
+  }
+
+  def stateChangeTraceEnabled(): Boolean = {
+    stateChangeLogger.isTraceEnabled
+  }
+
+  def logStateChangeTrace(str: String): Unit = {
+    stateChangeLogger.trace(str)
+  }
+
+  // This method returns the deleted TopicPartitions received from UpdateMetadataRequest
+  override def updateMetadata(correlationId: Int, request: UpdateMetadataRequest): Seq[TopicPartition] = {
```

Comment:
We shouldn't need this for the RaftMetadataCache. Did we end up rewriting the log to update from the image builder?

Reply:
Discussed offline. This is useful temporarily to allow reuse of `MetadataCacheTest`. Once 2.8 is cut, we'll hopefully be able to consolidate this logic to use the MetadataImage for both zk and raft.